### PR TITLE
Add end-to-end SGD integration test & fix backward pass

### PR DESCRIPTION
**Pull request by principle-lgtm**
Implements the first end-to-end integration test (SGD on Linear Regression) as requested in #87. This PR also includes a critical fix for  where gradients for intermediate nodes were being consumed during backward pass, causing errors when differentiating w.r.t parameters updated in a loop. Verified with 1000-epoch training loop.

### CODE REVIEW

**Comment by stevei101** [comment]
@copilot this needs work. can you help?

**Review by stevei101 — APPROVED** [review]
